README.md
# KMR_gamepads

Library for handling inputs from different gamepads, through the evdev interface.

Dependencies:
- libevdev

How to use: go to the KMR_gamepads folder, then:
```bash
mkdir build
cd build
cmake ../
cmake --build .
```

If you have Doxygen and Graphviz installed, you can regenerate the documentation locally with
```bash
make docs
```
from the `build` folder after the cmake.  
The generated documentation can be found in "docs/generated_docs/html".


CHANGELOG
# Change Log
All notable changes to this library will be documented in this file.

## [1.0.0] - 2023-22-09
First finished version of the library.

include/KMR_ps5.hpp
/**
 * KM-Robota library
 ******************************************************************************
 * @file            KMR_ps5.hpp
 * @brief           Header for the KMR_ps5.cpp file.
 ******************************************************************************
 * @copyright
 * Copyright 2021-2023 Laura Paez Coy and Kamilo Melo                    \n
 * This code is under MIT licence: https://opensource.org/licenses/MIT
 * @authors  Laura.Paez@KM-RoBota.com, 08/2023
 * @authors  Kamilo.Melo@KM-RoBota.com, 08/2023
 * @authors katarina.lichardova@km-robota.com, 08/2023
 ******************************************************************************
 */

#ifndef KMR_PS5_HPP
#define KMR_PS5_HPP

#include <iostream>
#include <mutex>
#include "libevdev-1.0/libevdev/libevdev.h"
#include <thread>


namespace KMR::gamepads
{

/**
 * @brief       Enumerate of all buttons present in a PS5
 * @details     The names follow the naming set by the evdev library:
 * https://gitlab.freedesktop.org/libevdev/libevdev/-/blob/master/include/linux/linux/input-event-codes.h 
 * @note        Even though HAT0X and HAT0Y are technically axes, they are classified as buttons
 *              given that they can only take 3 possible values {-1, 0, 1}
 */
enum ps5Buttons {e_BTN_SELECT, e_BTN_START, e_BTN_MODE, 
                 e_BTN_NORTH, e_BTN_WEST, e_BTN_SOUTH, e_BTN_EAST,
                 e_BTN_TR, e_BTN_TR2, e_BTN_TL, e_BTN_TL2,
                 e_BTN_THUMBR, e_BTN_THUMBL,
                 e_ABS_HAT0X, e_ABS_HAT0Y, 
                 NBR_BUTTONS, UNDEF_BUTTON};

/**
 * @brief       Enumerate of all axes present in a PS5
 */
enum ps5Axes{e_ABS_RX, e_ABS_RY, e_ABS_X, e_ABS_Y,
             e_ABS_Z, e_ABS_RZ,
             NBR_AXES, UNDEF_AXIS};

class PS5 {
    private:
        int m_fd; // File descriptor
        std::thread m_ps5_thread;

        void updateGamepad(input_event ev);
        void gamepadLoop(const char* gamepad_portname);

    public:
        int m_buttons[NBR_BUTTONS];
        float m_axes[NBR_AXES];
        std::mutex m_mutex;

        PS5();
        PS5(const char* gamepad_portname);
        void stop();
};

}

#endif

source/KM-R_ps5.cpp
/**
 ****************************************************************************
 * KM-Robota's ps5.cpp module
 ****************************************************************************
 * @file        KM-R_ps5.cpp
 * @brief       ps5
 ****************************************************************************
 * @copyright
 * Copyright 2021-2023 Laura Paez Coy and Kamilo Melo                    \n
 * This code is under MIT licence: https://opensource.org/licenses/MIT
 * @authors  Laura.Paez@KM-RoBota.com, 04-2023
 * @authors  Kamilo.Melo@KM-RoBota.com, 05-2023
 ****************************************************************************
 */

// Event types, directly defined in the linux kernel: 
// https://www.kernel.org/doc/Documentation/input/event-codes.txt
// Event codes, defined in libevdev: 
// https://gitlab.freedesktop.org/libevdev/libevdev/-/blob/master/include/linux/linux/input-event-codes.h

#include <iostream>
#include <fcntl.h>
#include <string.h>
#include <vector>

#include "KM-R_ps5.hpp"


using namespace std;

#define MAX_AXIS_VAL 255


/**
 * @brief       Initialize a PS5 object: to call after its constructor
 * @param[in]   gamepad_portname Port handling the gamepad, in the form of /dev/input/eventX
 * @retval      void
 */
void PS5::init(const char* gamepad_portname)
{
    struct libevdev *dev = NULL;
    int fd;
    int rc = 1;

    fd = open(gamepad_portname, O_RDONLY|O_NONBLOCK);
    m_fd = fd;

    rc = libevdev_new_from_fd(m_fd, &dev);
    if (rc < 0) {
        fprintf(stderr, "Failed to init libevdev (%s)\n", strerror(-rc));
        exit(1);
    }
    printf("Input device name: \"%s\"\n", libevdev_get_name(dev));
    printf("Input device ID: bus %#x vendor %#x product %#x\n",
        libevdev_get_id_bustype(dev),
        libevdev_get_id_vendor(dev),
        libevdev_get_id_product(dev));
}


/**
 * @brief       Read and save the current gamepad values. To loop in a thread
 * @retval      void
 */
void PS5::gamepadLoop()
{
    struct libevdev *dev = NULL;
    int rc = 1;
    int read_flag = LIBEVDEV_READ_FLAG_NORMAL;


    // Map buttons and axes
    do {
        struct input_event ev;
        rc = libevdev_next_event(dev, read_flag, &ev);
        if (rc == LIBEVDEV_READ_STATUS_SUCCESS)
            updateGamepad(ev);
        else if (rc == -EAGAIN)
            read_flag = LIBEVDEV_READ_FLAG_NORMAL;
        else if (rc == LIBEVDEV_READ_STATUS_SYNC){
            read_flag = LIBEVDEV_READ_FLAG_SYNC;
            updateGamepad(ev);
        }

    } while (1); 
}

/**
 * @brief       Save the new gamepad value into the correct table
 * @param[in]   ev Event indicating a change of gamepad value
 * @retval      void
 * @note        The values of the axes get remapped from [0,255] to [-1,1], 
 *              and the Y-axes inverted so that they point towards the front
 */
void PS5::updateGamepad(input_event ev)
{       
    int value = ev.value;     
    const char* event_type = libevdev_event_type_get_name(ev.type);
    const char* event_code = libevdev_event_code_get_name(ev.type, ev.code);

    // The value gets remapped from [0, 255] to [-1, 1] for axes
    float rescaled_value = 2/(float)MAX_AXIS_VAL * value -1;

    // Lock the mutex
    scoped_lock lock(m_mutex);

    // A button's state has changed
    if (strcmp(event_type, "EV_KEY") == 0){
        if      (strcmp(event_code, "BTN_SELECT") == 0) m_buttons[e_BTN_SELECT] = value;
        else if (strcmp(event_code, "BTN_START") == 0)  m_buttons[e_BTN_START] = value;
        else if (strcmp(event_code, "BTN_MODE") == 0)   m_buttons[e_BTN_MODE] = value;
        else if (strcmp(event_code, "BTN_NORTH") == 0)  m_buttons[e_BTN_NORTH] = value;
        else if (strcmp(event_code, "BTN_WEST") == 0)   m_buttons[e_BTN_WEST] = value;
        else if (strcmp(event_code, "BTN_SOUTH") == 0)  m_buttons[e_BTN_SOUTH] = value;
        else if (strcmp(event_code, "BTN_EAST") == 0)   m_buttons[e_BTN_EAST] = value;
        else if (strcmp(event_code, "BTN_TR") == 0)     m_buttons[e_BTN_TR] = value;
        else if (strcmp(event_code, "BTN_TR2") == 0)    m_buttons[e_BTN_TR2] = value;
        else if (strcmp(event_code, "BTN_TL") == 0)     m_buttons[e_BTN_TL] = value;
        else if (strcmp(event_code, "BTN_TL2") == 0)    m_buttons[e_BTN_TL2] = value;
        else if (strcmp(event_code, "BTN_THUMBR") == 0) m_buttons[e_BTN_THUMBR] = value;
        else if (strcmp(event_code, "BTN_THUMBL") == 0) m_buttons[e_BTN_THUMBL] = value;
    }

    // The value of an absolute axis has changed
    else if (strcmp(event_type, "EV_ABS") == 0){
        if     (strcmp(event_code, "ABS_RX") == 0)    m_axes[e_ABS_RX] =  rescaled_value;
        else if(strcmp(event_code, "ABS_RY") == 0)    m_axes[e_ABS_RY] = -rescaled_value;
        else if(strcmp(event_code, "ABS_X") == 0)     m_axes[e_ABS_X] =   rescaled_value;
        else if(strcmp(event_code, "ABS_Y") == 0)     m_axes[e_ABS_Y] =  -rescaled_value;
        else if(strcmp(event_code, "ABS_Z") == 0)     m_axes[e_ABS_Z] =   rescaled_value;
        else if(strcmp(event_code, "ABS_RZ") == 0)    m_axes[e_ABS_RZ] =  rescaled_value;
        else if(strcmp(event_code, "ABS_HAT0X") == 0) m_buttons[e_ABS_HAT0X] = value;
        else if(strcmp(event_code, "ABS_HAT0Y") == 0) m_buttons[e_ABS_HAT0Y] = -value;
    }
}


include/KM-R_ps5.hpp
/**
 * KM-Robota library
 ******************************************************************************
 * @file            KM-R_ps5.hpp
 * @brief           Header for the lib_robot.cpp file.
 ******************************************************************************
 * @copyright
 * Copyright 2021-2023 Laura Paez Coy and Kamilo Melo                    \n
 * This code is under MIT licence: https://opensource.org/licenses/MIT
 * @authors  Laura.Paez@KM-RoBota.com, 04-2023
 * @authors  Kamilo.Melo@KM-RoBota.com, 05-2023
 ******************************************************************************
 */

#ifndef KMR_PS5_HPP
#define KMR_PS5_HPP

#include <iostream>
#include <mutex>
#include "libevdev-1.0/libevdev/libevdev.h"

/**
 * @brief       Enumerate of all buttons present in a PS5
 * @details     The names follow the naming set by the evdev library:
 * https://gitlab.freedesktop.org/libevdev/libevdev/-/blob/master/include/linux/linux/input-event-codes.h 
 * @note        Even though HAT0X and HAT0Y are technically axes, they are classified as buttons
 *              given that they can only take 3 possible values {-1, 0, 1}
 */
enum ps5Buttons {e_BTN_SELECT, e_BTN_START, e_BTN_MODE, 
                 e_BTN_NORTH, e_BTN_WEST, e_BTN_SOUTH, e_BTN_EAST,
                 e_BTN_TR, e_BTN_TR2, e_BTN_TL, e_BTN_TL2,
                 e_BTN_THUMBR, e_BTN_THUMBL,
                 e_ABS_HAT0X, e_ABS_HAT0Y, 
                 NBR_BUTTONS, UNDEF_BUTTON};

/**
 * @brief       Enumerate of all axes present in a PS5
 */
enum ps5Axes{e_ABS_RX, e_ABS_RY, e_ABS_X, e_ABS_Y,
             e_ABS_Z, e_ABS_RZ,
             NBR_AXES, UNDEF_AXIS};

class PS5 {
    private:
        int m_fd; // File descriptor

        void updateGamepad(input_event ev);

    public:
        int m_buttons[NBR_BUTTONS];
        float m_axes[NBR_AXES];
        std::mutex m_mutex;

        void init(const char* gamepad_portname);
        void gamepadLoop();
};


#endif

source/KMR_ps5.cpp
/**
 ****************************************************************************
 * KM-Robota's ps5.cpp module
 ****************************************************************************
 * @file        KMR_ps5.cpp
 * @brief       ps5
 ****************************************************************************
 * @copyright
 * Copyright 2021-2023 Laura Paez Coy and Kamilo Melo                    \n
 * This code is under MIT licence: https://opensource.org/licenses/MIT
 * @authors  Laura.Paez@KM-RoBota.com, 08/2023
 * @authors  Kamilo.Melo@KM-RoBota.com, 08/2023
 * @authors katarina.lichardova@km-robota.com, 08/2023
 ****************************************************************************
 */

// Event types, directly defined in the linux kernel: 
// https://www.kernel.org/doc/Documentation/input/event-codes.txt
// Event codes, defined in libevdev: 
// https://gitlab.freedesktop.org/libevdev/libevdev/-/blob/master/include/linux/linux/input-event-codes.h

#include <iostream>
#include <fcntl.h>   // Control over the open file referenced by file descriptor 
#include <string.h> 

#include "KMR_ps5.hpp"


using namespace std;

#define MAX_AXIS_VAL 255

namespace KMR::gamepads
{

/**
 * @brief       Create and initialize a PS5 object
 * @retval      void
 * @note        By default, open Sony Dualsense gamepad. If you wish to use a different
 *              gamepad, use the overload of this constructor
 */
PS5::PS5()
{
    // Initialize tables
    for (int i=0; i<NBR_BUTTONS; i++)
        m_buttons[i] = 0;
    for (int i=0; i<NBR_AXES; i++)
        m_axes[i] = 0; 

    m_ps5_thread = thread(&KMR::gamepads::PS5::gamepadLoop, this,
    "/dev/input/by-id/usb-Sony_Interactive_Entertainment_Wireless_Controller-if03-event-joystick");

    cout << "Success: PS5 object created!" << endl;
}

/**
 * @brief       Create and initialize a PS5 object with a different gamepad than the default
 * @retval      void
 * @overload
 */
PS5::PS5(const char* gamepad_portname)
{
    // Initialize tables
    for (int i=0; i<NBR_BUTTONS; i++)
        m_buttons[i] = 0;
    for (int i=0; i<NBR_AXES; i++)
        m_axes[i] = 0; 

    m_ps5_thread = thread(&KMR::gamepads::PS5::gamepadLoop, this, gamepad_portname);

    cout << "Success: PS5 object created!" << endl;
}


/**
 * @brief       Read and save the current gamepad values. To loop in a thread
 * @param[in]   gamepad_portname Port handling the gamepad, in the form of /dev/input/eventX
 * @retval      void
 */
void PS5::gamepadLoop(const char* gamepad_portname)
{
    struct libevdev *dev = NULL;
    int fd = 1;
    int rc = 1;
    int read_flag = LIBEVDEV_READ_FLAG_NORMAL;

    fd = open(gamepad_portname, O_RDONLY|O_NONBLOCK);
    m_fd = fd;

    rc = libevdev_new_from_fd(m_fd, &dev);
    if (rc < 0) {
        fprintf(stderr, "Failed to init libevdev (%s)\n", strerror(-rc));
        exit(1);
    }
    printf("Input device name: \"%s\"\n", libevdev_get_name(dev));
    printf("Input device ID: bus %#x vendor %#x product %#x\n",
        libevdev_get_id_bustype(dev),
        libevdev_get_id_vendor(dev),
        libevdev_get_id_product(dev));


    // Start of loop
    do {
        struct input_event ev;
        rc = libevdev_next_event(dev, read_flag, &ev);
        if (rc == LIBEVDEV_READ_STATUS_SUCCESS)
            updateGamepad(ev);
        else if (rc == -EAGAIN)
            read_flag = LIBEVDEV_READ_FLAG_NORMAL;
        else if (rc == LIBEVDEV_READ_STATUS_SYNC){
            read_flag = LIBEVDEV_READ_FLAG_SYNC;
            updateGamepad(ev);
        }

    } while (1); 
}

/**
 * @brief       Save the new gamepad value into the correct table
 * @param[in]   ev Event indicating a change of gamepad value
 * @retval      void
 * @note        The values of the axes get remapped from [0,255] to [-1,1], 
 *              and the Y-axes inverted so that they point towards the front
 */
void PS5::updateGamepad(input_event ev)
{       
    int value = ev.value;     
    const char* event_type = libevdev_event_type_get_name(ev.type);
    const char* event_code = libevdev_event_code_get_name(ev.type, ev.code);

    // The value gets remapped from [0, 255] to [-1, 1] for axes
    float rescaled_value = 2/(float)MAX_AXIS_VAL * value -1;

    // Lock the mutex
    scoped_lock lock(m_mutex);

    // A button's state has changed
    if (strcmp(event_type, "EV_KEY") == 0){
        if      (strcmp(event_code, "BTN_SELECT") == 0) m_buttons[e_BTN_SELECT] = value;
        else if (strcmp(event_code, "BTN_START") == 0)  m_buttons[e_BTN_START] = value;
        else if (strcmp(event_code, "BTN_MODE") == 0)   m_buttons[e_BTN_MODE] = value;
        else if (strcmp(event_code, "BTN_NORTH") == 0)  m_buttons[e_BTN_NORTH] = value;
        else if (strcmp(event_code, "BTN_WEST") == 0)   m_buttons[e_BTN_WEST] = value;
        else if (strcmp(event_code, "BTN_SOUTH") == 0)  m_buttons[e_BTN_SOUTH] = value;
        else if (strcmp(event_code, "BTN_EAST") == 0)   m_buttons[e_BTN_EAST] = value;
        else if (strcmp(event_code, "BTN_TR") == 0)     m_buttons[e_BTN_TR] = value;
        else if (strcmp(event_code, "BTN_TR2") == 0)    m_buttons[e_BTN_TR2] = value;
        else if (strcmp(event_code, "BTN_TL") == 0)     m_buttons[e_BTN_TL] = value;
        else if (strcmp(event_code, "BTN_TL2") == 0)    m_buttons[e_BTN_TL2] = value;
        else if (strcmp(event_code, "BTN_THUMBR") == 0) m_buttons[e_BTN_THUMBR] = value;
        else if (strcmp(event_code, "BTN_THUMBL") == 0) m_buttons[e_BTN_THUMBL] = value;
    }

    // The value of an absolute axis has changed
    else if (strcmp(event_type, "EV_ABS") == 0){
        if     (strcmp(event_code, "ABS_RX") == 0)    m_axes[e_ABS_RX] =  rescaled_value;
        else if(strcmp(event_code, "ABS_RY") == 0)    m_axes[e_ABS_RY] = -rescaled_value;
        else if(strcmp(event_code, "ABS_X") == 0)     m_axes[e_ABS_X] =   rescaled_value;
        else if(strcmp(event_code, "ABS_Y") == 0)     m_axes[e_ABS_Y] =  -rescaled_value;
        else if(strcmp(event_code, "ABS_Z") == 0)     m_axes[e_ABS_Z] =   rescaled_value;
        else if(strcmp(event_code, "ABS_RZ") == 0)    m_axes[e_ABS_RZ] =  rescaled_value;
        else if(strcmp(event_code, "ABS_HAT0X") == 0) m_buttons[e_ABS_HAT0X] = value;
        else if(strcmp(event_code, "ABS_HAT0Y") == 0) m_buttons[e_ABS_HAT0Y] = -value;
    }
}

/**
 * @brief       Safely stops the gamepad's thread. To call at the end of the program
 */
void PS5::stop()
{
    m_ps5_thread.join();
}


}